### PR TITLE
[ValueTracking] Add more conditions in `isTruePredicate`

### PR DESCRIPTION
**Pull request by goldsteinn**
NB: This comes after #86082, I was adding in the `addlike` variants and just saw
the function was lacking in general.

There is one notable "regression". This patch replaces the bespoke `or
disjoint` logic with a direct match. This means we fail some
simplification during `instsimplify`.
All the cases we fail in `instsimplify` we do handle in `instcombine`
as we add `disjoint` flags.

Other than that, just some basic cases.

See proofs: https://alive2.llvm.org/ce/z/_-g7C8

### CODE REVIEW

File: llvm/test/Transforms/InstSimplify/implies.ll [hunk 1]
```diff
@@ -216,7 +222,13 @@ define i1 @test13(i32 %length.i, i32 %x) {
...
 define i1 @test14(i32 %length.i, i32 %x.full) {
 ; CHECK-LABEL: @test14(
-; CHECK-NEXT:    ret i1 true
+; CHECK-NEXT:    [[X:%.*]] = and i32 [[X_FULL:%.*]], -61681
+; CHECK-NEXT:    [[LARGE:%.*]] = or i32 [[X]], 8224
+; CHECK-NEXT:    [[SMALL:%.*]] = or i32 [[X]], 4112
+; CHECK-NEXT:    [[KNOWN:%.*]] = icmp ult i32 [[LARGE]], [[LENGTH_I:%.*]]
+; CHECK-NEXT:    [[TO_PROVE:%.*]] = icmp ult i32 [[SMALL]], [[LENGTH_I]]
+; CHECK-NEXT:    [[RES:%.*]] = icmp ule i1 [[KNOWN]], [[TO_PROVE]]
+; CHECK-NEXT:    ret i1 [[RES]]
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstSimplify/implies.ll:231]: These two regressions are b.c I removed the bespoke `or disjoint` detection logic. I figure its certainly more inline with our philosphy for these functions to rely on flags.
Note that if you run `InstCombine` (not just `InstSimplify`), these both get simplified to `true`.

**Comment by goldsteinn** [comment]
@nikic, IIRC you had started a patch on a helper function for value relationships and would be happy to drop this and pick that up. I couldn't find your patch though. Maybe I misremember?

**Comment by nikic** [comment]
@goldsteinn I think the patch you have in mind is https://github.com/llvm/llvm-project/pull/69471. Totally forgot about that one...

My general feedback on this patch is that we're re-implementing simplifyICmpInst() here. I think it would be good to at least check what the compile-time impact of using simplifyICmpInst() instead would be -- I suspect "bad", but best to confirm that...

Otherwise, I am fine with having some pragmatic duplication here, if having it is useful. I just want to make sure we don't end up with a full reimplementation down the line because a certain someone keeps spotting missing cases :)

File: llvm/test/Transforms/InstSimplify/implies.ll (hunk 1 above)
Comment (nikic) [reply llvm/test/Transforms/InstSimplify/implies.ll:231]: That's fine, but I'd adjust the test to add the disjoint flag.

**Comment by goldsteinn** [comment]
> > @goldsteinn I think the patch you have in mind is #69471. Totally forgot about that one...
> > My general feedback on this patch is that we're re-implementing simplifyICmpInst() here. I think it would be good to at least check what the compile-time impact of using simplifyICmpInst() instead would be -- I suspect "bad", but best to confirm that...
> > Otherwise, I am fine with having some pragmatic duplication here, if having it is useful. I just want to make sure we don't end up with a full reimplementation down the line because a certain someone keeps spotting missing cases :)
> 
> Let me look into creating a common helper for simlifyICmpInst and this.

I think impl is relatively novel, we have a scattering if the cases supported in
instsimplify like `simplifyICmpWithMinMax`, but the and/or icmp instructions
generally defer to is implied.

File: llvm/test/Transforms/InstSimplify/implies.ll (hunk 1 above)
Comment (goldsteinn) [reply llvm/test/Transforms/InstSimplify/implies.ll:244]: done

**Comment by goldsteinn** [comment]
ping.

**Comment by goldsteinn** [comment]
ping

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -8390,8 +8390,7 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
 
 /// Return true if "icmp Pred LHS RHS" is always true.
 static bool isTruePredicate(CmpInst::Predicate Pred, const Value *LHS,
-                            const Value *RHS, const DataLayout &DL,
-                            unsigned Depth) {
+                            const Value *RHS, const DataLayout &DL) {
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:8393]:
```suggestion
                            const Value *RHS) {
```
DL is unused.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request closed without merging**